### ISSUE TITLE
Failed to load native library 'libnative-platform.so' for Linux amd64.

### ISSUE THREAD

**Issue by ctmay4** [keeganwitt/docker-gradle#29]
I am running on a Linux hosted OpenShift container and trying to execute builds using Gitlab CI Runner.  I get the following error when using the Gradle Docker images:

```
Failed to load native library 'libnative-platform.so' for Linux amd64.
```

If I use the `gradle:4.2.0-jdk8` I get the error.  If I use `image: java:8` I do not get the error.  I tried adding the stacktrace and info options but it fails right away before supplying any help.

I saw there a similar issue that was closed but he said it had to do with OS X so this seems different.

```
Running with gitlab-ci-multi-runner 9.0.2 (fa8b86d)
  on gitlab-runner-1-bv0xl (53cb5409)
Using Kubernetes namespace: gitlab-runner
Using Kubernetes executor with image gradle:4.2.0-jdk8 ...
Waiting for pod gitlab-runner/runner-53cb5409-project-586-concurrent-0w41s5 to be running, status is Pending
Waiting for pod gitlab-runner/runner-53cb5409-project-586-concurrent-0w41s5 to be running, status is Pending
Waiting for pod gitlab-runner/runner-53cb5409-project-586-concurrent-0w41s5 to be running, status is Pending
Waiting for pod gitlab-runner/runner-53cb5409-project-586-concurrent-0w41s5 to be running, status is Pending
Waiting for pod gitlab-runner/runner-53cb5409-project-586-concurrent-0w41s5 to be running, status is Pending
Waiting for pod gitlab-runner/runner-53cb5409-project-586-concurrent-0w41s5 to be running, status is Pending
Waiting for pod gitlab-runner/runner-53cb5409-project-586-concurrent-0w41s5 to be running, status is Pending
Waiting for pod gitlab-runner/runner-53cb5409-project-586-concurrent-0w41s5 to be running, status is Pending
Waiting for pod gitlab-runner/runner-53cb5409-project-586-concurrent-0w41s5 to be running, status is Pending
Waiting for pod gitlab-runner/runner-53cb5409-project-586-concurrent-0w41s5 to be running, status is Pending
Waiting for pod gitlab-runner/runner-53cb5409-project-586-concurrent-0w41s5 to be running, status is Pending
Waiting for pod gitlab-runner/runner-53cb5409-project-586-concurrent-0w41s5 to be running, status is Pending
Waiting for pod gitlab-runner/runner-53cb5409-project-586-concurrent-0w41s5 to be running, status is Pending
Waiting for pod gitlab-runner/runner-53cb5409-project-586-concurrent-0w41s5 to be running, status is Pending
Waiting for pod gitlab-runner/runner-53cb5409-project-586-concurrent-0w41s5 to be running, status is Pending
Waiting for pod gitlab-runner/runner-53cb5409-project-586-concurrent-0w41s5 to be running, status is Pending
Waiting for pod gitlab-runner/runner-53cb5409-project-586-concurrent-0w41s5 to be running, status is Pending
Waiting for pod gitlab-runner/runner-53cb5409-project-586-concurrent-0w41s5 to be running, status is Pending
Running on runner-53cb5409-project-586-concurrent-0w41s5 via gitlab-runner-1-bv0xl...
Cloning repository...
Cloning into '/may/algorithms'...
Checking out 262fdd32 as master...
Skipping Git submodules setup
Checking cache for master...
Successfully extracted cache
$ export GRADLE_USER_HOME=`pwd`/.gradle
$ gradle --build-cache --stacktrace --info assemble

FAILURE: Build failed with an exception.

* What went wrong:
Failed to load native library 'libnative-platform.so' for Linux amd64.

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output.

* Get more help at https://help.gradle.org
ERROR: Job failed: error executing remote command: command terminated with non-zero exit code: Error executing in Docker Container: 1
```

**Comment by ron-reaser** [comment]
I also have this issue. I believe it was fixed in Gradle 4.2.1, and the current Gradle image for Docker only goes up to 4.2.0, so once the image is updated it should be resolved.

**Comment by keeganwitt** [comment]
I've opened a [PR](https://github.com/docker-library/official-images/pull/3552) with Docker library to update to 4.2.1.  Let me know once it's merged if it doesn't resolve your problem.

**Comment by ron-reaser** [comment]
I have tested gradle:alpine, gradle:4.2.1-jdk-alpine, and gradle:4.2.1-jdk and the failure to load the native library is not resolved.

**Comment by keeganwitt** [comment]
What user are you running in the container?  And do you have the Gradle home volume mounted?

**Comment by ron-reaser** [comment]
The image is being run by the latest version of GitLab Runner (CI). I believe it's run as an unprivileged user. It does have access to the Gradle home volume, as far as I know. For reference, if I use frekele/gradle from Docker instead of library/gradle, the problem does not occur.

**Comment by keeganwitt** [comment]
So the only times I've seen this is the past is if your volume isn't writable by the uid the container runs under (1000).  I'd suggest trying running the container with `-u root`.

**Comment by ron-reaser** [comment]
For my situation, running as root is not a safe fix, but I believe the volume is writable anyway . Here's all the relevant information I've read: [1](https://discuss.gradle.org/t/started-getting-libnative-platform-so-error-running-as-non-root-user/15593), [2](https://github.com/gradle/gradle/issues/1493).

**Comment by keeganwitt** [comment]
FYI, if I'm right about the volume permission being the issue, [frekele/gradle](https://hub.docker.com/r/frekele/gradle/) wouldn't be affected as it runs as root.  Is this a Java project or native project?

**Comment by ron-reaser** [comment]
Using the Docker executor in GitLab Runner to run Gradle to compile a Java project. The only thing I change is whether I use frekele/gradle or library/gradle, so I think that means the privilege scenario (root or not) would be the same for both, at least from outside the container (not counting anything that happens in the Dockerfile). I'm not really experienced enough with Docker or Dockerfiles to be able to help much, but I know that everything from outside the container (in terms of how it is executed) is handled identically regardless of the image. The only variable could be what happens inside the container. If you mean that's where the privilege issue occurs, I'm afraid that's beyond me.

**Comment by keeganwitt** [comment]
Yep, that's what I mean; the user that it's running as inside the container.  So if you change the command to be `docker run -u root ...` it'll run as root inside the container instead of the default user I created in this container (so people can avoid using root when they can).

**Comment by ron-reaser** [comment]
So does that mean you're confirming the frekele/gradle _does_ run as root inside the container, but library/gradle does not, and that's the cause of the problem? If so, I'll try that fix.

**Comment by keeganwitt** [comment]
Yes.  Exactly.
```$ docker run --rm -it frekele/gradle whoami
...
root
...```

**Comment by ron-reaser** [comment]
That makes sense. I can't confirm or deny a fix, because in my particular scenario I can't test any of the known fixes, but I appreciate the information.

**Comment by keeganwitt** [comment]
I'm not familiar with Concourse, but you could temporarily add a Bash step to your job if you wanted, right?

At the very least you should be able to change the container user, right?  Concourse is an important use case I think, so I just wanted to be sure that this image wasn't unusable on that platform.

**Comment by ron-reaser** [comment]
I don't know what Concourse is. Another CI service? The OP and myself have an issue with GitLab Runner as it interacts with the Docker Gradle image. It's unusable on that platform, but that's probably a niche concern, and the fix would be for me to make a Dockerfile which pulls from your image and just replaces the user directive.

**Comment by keeganwitt** [comment]
Oops, I was thinking of another issue.

Hmm, the options for the Docker Executor in Gitlab are pretty limited (for example, you also can't map ports).  The only ways around that from what I've read is Docker in Docker (which has some privilege escalation risks) or use the Shell Executor with the Docker commands in a script.

Alternatively, you could add your own image to run (and upload to your own repo on Dockerhub) like this
```
FROM gradle:alpine
USER root
```

If this is a problem in several places, I guess I'll bring back root, but I'm generally inclined to avoid it if I can.

**Comment by keeganwitt** [comment]
I just uploaded images with the root Dockerfile change, if you'd like to give that a try: https://hub.docker.com/r/keeganwitt/docker-gradle-root/.

**Comment by ron-reaser** [comment]
Yes, the Dockerfile you listed is the solution I was thinking would work, so the confirmation is helpful (I haven't tried it). It's probably the solution I'll eventually adopt, because I need other stuff installed inside the image anyway, and then I'll have to put it in the GitLab Registry, which is a whole 'nother thing.

Thank you a lot for your help. I've done a lot of searching on this issue, and this is the first time anybody on the internet has bothered to explain what the actual problem is. So hopefully this issue and your posted solutions will help others with the same problem in posterity.

**Comment by bradjones1** [comment]
It may also be sufficient to set the `GRADLE_USER_HOME` environment variable to a directory writable by the user.

**Comment by ron-reaser** [comment]
Unfortunately, I already do that, and it creates the directory as the root user. I'm not sure why. So then when I go to access it, I still have the problem.

**Comment by bradjones1** [comment]
@ron-reaser You could always specify the run-as user inside the container with `docker run -u ``id -u`` ...` ?

**Comment by ron-reaser** [comment]
My use case does not allow that (see previous comments).

**Comment by keeganwitt** [comment]
The problem (at least in his case) is the volume containing project he's building not being writable, not the Gradle home.

**Comment by razvan-panda** [comment]
I am getting the same error on Gradle 4.5 on NixOS

**Comment by keeganwitt** [comment]
@razvan-panda You're a GitLab user as well?  Has GitLab commented at all on this issue?

**Comment by razvan-panda** [comment]
@keeganwitt I am not a GitLab user. But nothing I could find helped me fix the issue, especially since it built successfully from the terminal and it only fails during the nix-build.

More info here: https://stackoverflow.com/q/48652914/750216

**Comment by keeganwitt** [comment]
This looks like you're running Gradle itself, not the Docker Gradle image.  Have I misunderstood?

**Comment by razvan-panda** [comment]
@keeganwitt I am having this issue when running Gradle via Nix and not using the Docker Gradle image.
But mentioning that the issue happens with Gradle 4.5 also might be useful for this ticket.

**Comment by Labi0** [comment]
I got this issue with gradle 4.2.1 and RHEL 7.4; Using some info provided here, I discovered that the user executing did not have write permission to create the gradle-user-home in the parent-directory. Creating the gradle-user-home directory with the desired permission resolved the issue.

**Comment by ebuildy** [comment]
Same issue because I changed the user to match GitlabCI ``gitlab`` user, I managed to fix it, here the Docker command I run:

```
#!/bin/sh

BUILD_CMD="gradle --info build"

if test -n "$USER_UID" && test -n "$USER_NAME"; then
    set -eux
    adduser --uid $USER_UID --quiet --home /home/gradle --disabled-password --gecos "djobi" $USER_NAME
    chown -R $USER_UID /home/gradle
    exec su -c "$BUILD_CMD" $USER_NAME
else
    set -eux
    exec $BUILD_CMD
fi

```

.gitlab-ci.yml

```
build java:
  stage: build
  script:
  - docker-compose run -e USER_NAME=$USER -e USER_UID=$UID --rm builder
```

**Comment by Beau-1983** [comment]
when I  use a normal user (not root user) to run gradle -v in my ubuntu system ,it outprint "Failed to load native library 'libnative-platform.so' for Linux amd64.",but root user can work.

here is the logs:


beau@BeauPC:/opt$ sudo su -
[sudo] beau 的密码： 
root@BeauPC:~# gradle -v

------------------------------------------------------------
Gradle 5.1-rc-2
------------------------------------------------------------

Build time:   2018-12-17 22:42:01 UTC
Revision:     4f93941ff2f23ed981b9416a66af36ab0a8ef4de

Kotlin DSL:   1.1.0
Kotlin:       1.3.11
Groovy:       2.5.4
Ant:          Apache Ant(TM) version 1.9.13 compiled on July 10 2018
JVM:          1.8.0_181 (Oracle Corporation 25.181-b13)
OS:           Linux 4.15.0-43-generic amd64

root@BeauPC:~# exit 
注销
beau@BeauPC:/opt$ gradle -v

FAILURE: Build failed with an exception.

* What went wrong:
Failed to load native library 'libnative-platform.so' for Linux amd64.

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org




-------------------

Is anyone who can help me to resolve this problem ?  please send me email to 462620105@qq.com,ths!

**Comment by kphannan** [comment]
After some experimentation and internet spelunking I found that the problem is permissions on the directory where libnative-platform.so is extracted to from the appropriate jar file(s) that comprise gradle.  This directory needs to be writable by the user running gradle.    An easy way to fix the issue is to either set GRADLE_USER_HOME to some directory writable by the user running gradle; or specify the directory on the command line `gradle -g $(pwd)`
